src/mongo/registerMongoCommands.ts

```diff
@@ -94,29 +98,49 @@ export function registerMongoCommands(context: vscode.ExtensionContext, actionHa
     });
     actionHandler.registerCommand('cosmosDB.launchMongoShell', launchMongoShell);
     actionHandler.registerCommand('cosmosDB.newMongoScrapbook', async () => await vscodeUtil.showNewFile('', context.extensionPath, 'Scrapbook', '.mongo'));
-    actionHandler.registerCommand('cosmosDB.executeMongoCommand', async function (this: IActionContext) {
+    actionHandler.registerCommand('cosmosDB.executeMongoCommand', async function (this: IActionContext, commandText: object) {
+        await loadPersistedMongoDBTask;
+        if (typeof commandText === "string") {
+            await MongoCommands.executeCommandFromText(<IAzureParentNode<MongoDatabaseTreeItem>>ext.connectedMongoDB, context.extensionPath, editorManager, tree, this, <string>commandText);
+        } else {
+            await MongoCommands.executeCommandFromActiveEditor(<IAzureParentNode<MongoDatabaseTreeItem>>ext.connectedMongoDB, context.extensionPath, editorManager, tree, this);
+        }
+    });
+    actionHandler.registerCommand('cosmosDB.executeAllMongoCommands', async function (this: IActionContext) {
         await loadPersistedMongoDBTask;
-        await MongoCommands.executeCommandFromActiveEditor(<IAzureParentNode<MongoDatabaseTreeItem>>ext.connectedMongoDB, context.extensionPath, editorManager, tree, this);
+        await MongoCommands.executeAllCommandsFromActiveEditor(<IAzureParentNode<MongoDatabaseTreeItem>>ext.connectedMongoDB, context.extensionPath, editorManager, tree, this);
     });
 }
 
-async function loadPersistedMongoDB(context: vscode.ExtensionContext, tree: AzureTreeDataProvider, languageClient: MongoDBLanguageClient): Promise<void> {
+async function loadPersistedMongoDB(context: vscode.ExtensionContext, tree: AzureTreeDataProvider, languageClient: MongoDBLanguageClient, codeLensProvider: MongoCodeLensProvider): Promise<void> {
+    let persistedNode: IAzureNode | undefined;
+
     await callWithTelemetryAndErrorHandling('cosmosDB.loadPersistedMongoDB', reporter, undefined, async function (this: IActionContext): Promise<void> {
         this.suppressErrorDisplay = true;
         this.properties.isActivationEvent = 'true';
         const persistedNodeId: string | undefined = context.globalState.get(connectedDBKey);
         if (persistedNodeId) {
-            const persistedNode: IAzureNode | undefined = await tree.findNode(persistedNodeId);
-            if (persistedNode) {
-                await languageClient.client.onReady();
-                await vscode.commands.executeCommand('cosmosDB.connectMongoDB', persistedNode);
-            }
+            persistedNode = await tree.findNode(persistedNodeId);
         }
     });
+
+    if (persistedNode) {
```

Comment:
Why did you move this code outside of `callWithTelemetryAndErrorHandling`? It was put in there since we await 'loadPersistedMongoDBTask' multiple times and we don't want the error being displayed multiple times

Reply:
Doesn't matter how many times it's awaited, it will only run once.  

CodeLens will show "Initializing" until told setConnectedDatabase called first time, so have to make sure that happens. However, current version doesn't handle errors connecting to DB, I think this next one will be better...

Reply:
If you await a promise multiple times, you're correct that it only runs once. However, if the result of awaiting the promise is an error, you will get that error thrown every single time you await it. For example you can test this out by putting `throw new Error('Hi Stephen!');` right before `codeLensProvider.setConnectedDatabase(undefined);`. The error will pop up every time you execute a mongo command (not just the first time).

Instead, I suggest moving everything inside `await callWithTelemetryAndErrorHandling(` and using a 'try/finally' to make sure CodeLens always gets initialized

Reply:
setConnectedDatabase can't fail, but yeah I agree it's better inside the telemetry.

Reply:
"can't fail" <- famous last words 😂

Reply:
"As currently implemented" (tm)